src/memray/_memray.pyx

```diff
@@ -318,97 +321,130 @@ def start_thread_trace(frame, event, arg):
+        except OSError as exc:
+            raise OSError(f"Could not open file {file_name}: {exc.strerror}") from None
 
-    cdef void _populate_allocations(self):
-        cdef RecordReader* reader = self._get_reader()
+        self._path = "/proc/self/fd/" + str(self._file.fileno())
+        self._populate_metadata()
 
+    def __dealloc__(self):
+        self.close()
+
+    cdef object _populate_metadata(self):
+        """Initial pass to populate _header, _high_watermark, and _memory_records."""
+        cdef HighWatermarkFinder finder
+        cdef shared_ptr[RecordReader] reader_sp = make_shared[RecordReader](
+            unique_ptr[FileSource](new FileSource(self._path)),
+            False
+        )
+        cdef RecordReader* reader = reader_sp.get()
+
+        self._header = reader.getHeader()
         stats = self._header["stats"]
-        self._allocations.reserve(stats["n_allocations"])
+
         n_memory_records_approx = 2048
         if 0 < stats["start_time"] < stats["end_time"]:
             n_memory_records_approx = (stats["end_time"] - stats["start_time"]) / 10
         self._memory_records.reserve(n_memory_records_approx)
 
         while True:
+            PyErr_CheckSignals()
             ret = reader.nextRecord()
             if ret == RecordResult.RecordResultAllocationRecord:
-                self._allocations.push_back(reader.getLatestAllocation())
+                finder.processAllocation(reader.getLatestAllocation())
             elif ret == RecordResult.RecordResultMemoryRecord:
                 self._memory_records.push_back(reader.getLatestMemoryRecord())
             else:
                 break
 
-    cdef RecordReader* _get_reader(self) except *:
-        if self._reader.get() == NULL:
-            raise ValueError("Operation on a closed FileReader")
-        return self._reader.get()
+        self._high_watermark = finder.getHighWatermark()
 
-    def close(self):
-        cdef RecordReader* reader = self._get_reader()
-        reader.close()
+    cpdef close(self):
+        if self._file is not None:
+            file = self._file
+            self._file = None
+            file.close()
 
-    cdef void _ensure_reader_is_open(self) except *:
-        if not self._get_reader().isOpen():
+    cdef void _ensure_not_closed(self) except *:
+        if self._file is None:
             raise ValueError("Operation on a closed FileReader")
 
     @property
     def closed(self):
-        return not self._get_reader().isOpen()
+        return self._file is None
 
     def __enter__(self):
         return self
 
     def __exit__(self, exc_type, exc_value, exc_traceback):
         self.close()
 
-    def __dealloc__(self):
-        self._reader.reset()
-    
-    def _yield_allocations(self, size_t index, merge_threads):
-        for elem in Py_GetSnapshotAllocationRecords(
-            self._allocations, index, merge_threads):
+    def _yield_allocations(self, size_t max_records, bool merge_threads):
```

Comment:
Let's rename this as this now is *agregating* the allocations (so technically cannot be used also in the `get_allocation_records()`.

Reply:
This was already aggregating the allocations, and already wasn't used in `get_allocation_records()`. I didn't change anything about that.

I'm not sure what a better name would be. I'll change it to
```py
def _yield_unfreed_allocations(self, size_t allocations_to_process, bool merge_threads):
```

I don't think that's perfect, but I'm struggling to come up with anything better. If you've got something, feel free to fixup.